Remove `Denomination::MilliSatoshi`

jamillambert opened the pull request:
The denomination `MilliSatoshi` has been removed from `Amount`. 

`NanoBitcoin` and `PicoBitcoin` have also been removed since they are also less than 1 satoshi and the same reason for removing milliSatoshi in Issue #2820 should apply to them.

The second patch fixes the way the denominations with various combinations of upper and lower case are handled when converted from a string.

Close: #2820

tcharding reviewed (commented):
We probably should ask _why_ @Kixunil implemented all this originally, he doesn't usually do stuff for no reason. (Mentioning him to see if he responds.)

apoelstra commented:
Agreed let's give Kix a week or so to notice and respond to this. But for my part, ACK 749af0011ddaac866267c515b925eacdaf68aa82. I remember some of the original discussion, wherein we wanted simultaneously to be case-insensitive (why? I dunno. But I dunno why we wanted to support these units in the first place. It seemed harmless enough) and also to avoid ambiguity between Peta and pico, Mega and milli, etc.

So I believe the logic in this PR is right.

**But** if it were up to me, we would stop trying to be case-insensitive, and instead accept only correctly-cased SI prefixes followed by `BTC` or `btc`.

Kixunil commented on units/src/amount.rs:
```diff
@@ -88,37 +76,26 @@ impl Denomination {
-            "SATOSHI" | "satoshi" | "SATOSHIS" | "satoshis" | "SAT" | "sat" | "SATS" | "sats" =>
-                Some(Denomination::Satoshi),
-            "mSAT" | "msat" | "mSATs" | "msats" => Some(Denomination::MilliSatoshi),
+        let s_lower = s.to_lowercase();
```
Yes, I think we should keep it.

Kixunil commented on units/src/amount.rs:
```diff
@@ -73,11 +64,8 @@ impl Denomination {
-            Denomination::PicoBitcoin => 4,
             Denomination::Bit => -2,
             Denomination::Satoshi => 0,
-            Denomination::MilliSatoshi => 3,
```
To better convey that we no longer support sub-satoshi units we should return `u8` from `precision`. I suspect it'll clean up some other code as well.

Kixunil reviewed (commented):
I think we should change the comment claiming mixed case is OK, not the code. Disallowing stuff like `mBtC` is desirable.

I think the code was written to still return an error even for `MsAt` - it'd just return a different one. So we shouldn't worry about it too much. If you want to have great error messages we need multiple errors. I started work on it a long time ago but it was complicated and I didn't have time to finish.

Kixunil commented:
I'm not sure what the question to me *specifically* is. I believe I didn't decide whether to add msat or not, I only wanted it to be consistent. Two days ago we had rust-bitcoin dinner at BTC Prague and we agreed that msat is weird here and should probably be removed. *However* I think we should expose a way to format msat using our formatting code since the code is pretty complicated, it'll be useful. We could do the same for parsing (and don't have the precision limitation).

apoelstra commented:
> _However_ I think we should expose a way to format msat using our formatting code since the code is pretty complicated, it'll be useful.

Agreed. Though hopefully this isn't too annoying. Maybe we will need a separate method for it. I don't want to split the `Denomination` enum into two, and I don't want a `Denomination::MilliSatoshi` variant that can't be used for parsing either.

Kixunil commented:
I was thinking of having just free-standing functions using `u64`.

jamillambert commented:
I have changed the patches to now only accept all upper case or all lower case denominations, and where `BTC` or `btc` can be prefixed with a lower case c, m or u.  And changed the comment to state this.  

I also put back the const array and changed it to only contain the cases where the denomination has a capital prefix but is otherwise valid.

I wasn't really sure what was meant by having a separate function for formatting `msat`, without supporting the denomination `msat`, so have not addressed this.  But I guess this can be dealt with as a separate issue.

apoelstra reviewed (approved):
ACK 9ea8c58ad6f077a8ff07a7ccec458959b120f075 though if you have to edit the PR again, maybe add a comment saying that a MBTC is not a realistic amount of bitcoin and a Msat is the same as a cBTC. So people do not think we are unreasonable excluding M.

Kixunil commented on units/src/amount.rs:
```diff
@@ -98,8 +98,8 @@ impl Denomination {
-const CONFUSING_FORMS: [&str; 9] =
-    ["Msat", "Msats", "MSAT", "MSATS", "MSat", "MSats", "MBTC", "Mbtc", "PBTC"];
+const CONFUSING_FORMS: [&str; 6] =
+    ["MBTC", "Mbtc", "CBTC", "Cbtc", "UBTC", "Ubtc"];
```
Strictly speaking C and U are not confusing but I agree with banning them because they look too weird. :)

Kixunil reviewed (dismissed):
ACK 9ea8c58ad6f077a8ff07a7ccec458959b120f075

Kixunil commented:
> Msat is the same as a cBTC

FYI I find Msat more intuitive than cBTC and I actually do use it very often in real text in particular when talking about LN channel sizes since they often are in the order of (ten) million sats. But I prefer parsing to throw an error which is easily fixed rather than someone losing satoshi.

Also I've knowingly ACKed a commit with a broken CI check because I believe that committing information to git that can be deterministically computed from the source code is a bad practice and we need a different solution to whatever it's trying to achieve.

apoelstra commented:
> we need a different solution to whatever it's trying to achieve

What it's trying to achieve is having a permanent and easily-viewable record of changes to the API. Putting the API into git seems like exactly the right solution to this.  (Ok, I guess, writing some external git extension would be morally better but practically quite annoying.)

Kixunil commented:
Do we really need to know when exactly some API changed? Isn't per-version record enough?

apoelstra commented:
>Do we really need to know when exactly some API changed? Isn't per-version record enough?

No, a per-version record would be completely unreadable. It woud be overwhelmed by diff for minor/harmless changes (e.g. adding `const` to functions) and for deliberate-but-large changes (e.g. accepting keys and hashes by value rather than by reference) and it would be impossible to see meaningful and unintentional/surprising changes.

And even if we could see changes that we cared about, we would then need to translate the API diff into a source code diff, use `git log -S` to find the relevant commit, and use some commit-to-PR index (I have one but I don't think most people do) to find the revelant PR. And then we'd have to go comment on that long-merged PR trying to re-open a discussion about an API change that happened maybe many months ago.

Meanwhile, real-time PRs would be harder to review because the API changes would be implicit in the code changes.

tcharding commented:
Anecdotally the api files have already, multiple times, made me catch mistakes in my own work, and have made reviewing other peoples patches easier (faster to do, and faster to check they are sane).

tcharding commented:
dev work flow is annoying though because of trailing api change patch but I think its worth it.

Kixunil commented:
> Meanwhile, real-time PRs would be harder to review because the API changes would be implicit in the code changes.

Can't we just have a bot posting the diff in the comments? And possibly instead of CI check have a post-merge action to commit the API if that's really needed. But I'd still prefer figuring out some out-of-git index - it also has the advantage that we get to know the API changes from before the tool was introduced.

Also if we do out-of-git tool, we could get rid of `diff` since it clutters the review with unchanged stuff. For API changes we probably only need to know "this item was removed" and "this item was added". This is trivial to code - just have two sets and compute their difference both ways. (Union minus intersect, but have to remember which item came from which set.)

> Anecdotally the api files have already, multiple times, made me catch mistakes in my own work

Anecdotally, I also became not hungry after eating french fries, but that doesn't mean they were the best option. :)

apoelstra commented:
> Can't we just have a bot posting the diff in the comments? And possibly instead of CI check have a post-merge action to commit the API if that's really needed.

No. Even when Github preserves data and when we are working online with an open browser and can access that data, mapping git history to Github PRs is difficult and locating specific comments within threads is difficult.

>But I'd still prefer figuring out some out-of-git index - it also has the advantage that we get to know the API changes from before the tool was introduced.
> 

Sure, I'd be happy with this. But AFAIK no such tool exists.

> Also if we do out-of-git tool, we could get rid of `diff` since it clutters the review with unchanged stuff. For API changes we probably only need to know "this item was removed" and "this item was added". This is trivial to code - just have two sets and compute their difference both ways. (Union minus intersect, but have to remember which item came from which set.)

FWIW you can use `diff -U0` to show only changes, no context.

jamillambert commented:
Added [api: Run just check-api](https://github.com/rust-bitcoin/rust-bitcoin/pull/2870/commits/362c3b8772fe26aa6c84e8e1fd3d9cc6cb5384cb) to fix CI fail

Kixunil reviewed (approved):
ACK 362c3b8772fe26aa6c84e8e1fd3d9cc6cb5384cb

tcharding reviewed (approved):
ACK 362c3b8772fe26aa6c84e8e1fd3d9cc6cb5384cb
